sdks/python/apache_beam/runners/interactive/caching/expression_cache.py

```diff
@@ -0,0 +1,90 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import *
+
+import apache_beam as beam
+from apache_beam.dataframe import convert
+from apache_beam.dataframe import expressions
+
+
+class ExpressionCache(object):
```

Comment:
Is the `ExpressionCache` to be used anywhere besides the `deferred_df_to_pcollection`? Can `replace_with_cached` be part of the constructor? If not, can we document how to use `ExpressionCache`? Like the lifecycle of a ExpressionCache. When to create a new `ExpressionCache`? When to invoke `replace_with_cached`? What are the side effects to the given expression and the interactive environment? How long does the side effects last?

I'm wondering how the cache is persisted between multiple runs? Or is it just a temporary helper class to alter the given expression.

Reply:
Added a comment to answer your questions here. My personal style is to not have constructors do too much work, so I kept the `replace_with_cached` where it is. The lifecycle is temporary and the only side-effects are in the modified expression. The cache isn't persisted between multiple runs, it just uses existing caches from the dataframe and interactive modules.